### PR TITLE
Phase 1 Pixel DQM (number 3, final?)

### PR DESCRIPTION
**Pull request by schneiml**
This is a follow up to #14015 and #13798 .

All commits are rebased to the CMSSW_8_1_CLANG_X_2016-04-24-0000 IB such that #13987 lies below, with fixes to make things actually work.

Some points mentioned in #14015 are still relevant, but everything that I got feedback to is fixed.

@makortel regarding the Certification module, I now remove that in DQMOffline_CRT_cff.py, even though I still don't get how that finds its way into the offline config (cert-stuff is removed in SiPixelOfflineDQM). And, said module crashes due to code like this `float dcsFrac = (iGetter.get("Pixel/EventInfo/DCSSummary"))->getFloatValue();`, though I can't see immediatly which histogram is missing due to my changes (probably all...).

@slava77 I now make the TrackerTopology internal structs visible to the outside. I am not sure whether that is a good idea, but I agree it is better than what we had before.

@smuzaffar there are now 3 new packages, one with shared code and two that use this code for digis and clusters. More packages will come in the future, but I can't tell yet how we will call them.
UPDATE: now 4. Two for shared code and config, two for the plugins.

Overall, I think this is in a mergeable shape now. Let me know if you think different. Also, chasing IBs and rebasing ~100 commits each time starts to get annoying.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by schneiml** [comment]
I cannot reproduce these problems after a clean

    cmsrel CMSSW_8_1_X_2016-04-24-0000
    cd CMSSW_8_1_X_2016-04-24-0000/src
    cmsenv
    git cms-merge-topic 14230
    scram b -j4 # takes forever since TrackerTopology is touched
    runTheMatrix.py -l 1325.0 # etc.

I will however push soon a fix for the static analyzer warnings.

Judging by the name `frontierHttpClnt_usinglastproxyingroup` where the tests crash, I'd guess it is a problem on the build machine. (Also, I wonder where the source code for that lives?)

**Comment by deguio** [comment]
please test

**Comment by deguio** [comment]
as discussed at todays meeting, the usage of th1f is suggested in place of the th1d histos
apart from that the PR looks good and can go in. more development will come on top for the online DQM.

File: DQM/SiPixelPhase1Common/python/MessageLogger_cfi.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import FWCore.ParameterSet.Config as cms
+
+MessageLogger = cms.Service(
```
Comment (slava77) [inline DQM/SiPixelPhase1Common/python/MessageLogger_cfi.py]: this overrides the central definition of MessageLogger when included in standard configuration.
This should not be done.

**Comment by slava77** [comment]
There is a new warning showing up at beginrun, apparently in all workflows (run1, run2, 2017)
```
%MSG-w SiStripCabling:   SiStripMonitorCluster:SiStripCalZeroBiasMonitorCluster@streamBeginRun  26-Apr-2016 15:48:16 CEST Run: 165121
[FedChannelConnection::apvPairNumber] Unexpected number of APV pairs: 65535
```
this has to be fixed (use eras to configure correctly) or silenced.

File: DataFormats/TrackerCommon/interface/TrackerTopology.h [hunk 2]
```diff
@@ -576,7 +576,10 @@ class TrackerTopology {
   std::string print(DetId detid) const;
 
   SiStripDetId::ModuleGeometry moduleGeometry(const DetId &id) const; 
-
+ 
+  PixelBarrelValues const& getPBVals() const { return pbVals_; };
+  PixelEndcapValues const& getPFVals() const { return pfVals_; };
```
Comment (slava77) [inline DataFormats/TrackerCommon/interface/TrackerTopology.h]: since you need to update in other files, please remove  semicolons at the end of the two lines above

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
the last test failed with ``` socket.timeout: timed out ```
and the currently available IB is broken

**Comment by slava77** [comment]
@cmsbuild please test
... apparently, tests that were started more recently now complete

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14230	3c45939
- changes in reco are rather minimal, exposing internal structure of TrackerTopology in PixelBarrelValues and PixelEndcapValues, needed for DQM organized partly using this internal layout.
- jenkins tests pass and comparisons with baseline show no difference (compared to the last checked commits as of 29fc5ce, the log files are now clean)

**Comment by dmitrijus** [comment]
+1

**Comment by schneiml** [comment]
@davidlange6 can we get the new packages added and this PR merged?

**Comment by fioriNTU** [comment]
@davidlange6 @slava77 the Phase I DQM is a fundamental tool for us, we need to test it extensively in the Tracker DQM group, use it to validate the new geometry and relvals. It would be very important to have this PR merged as soon as possible. It would also ease the share of work in our group. Do you have a time estimate to have it done, and possibly to have a prerelease with this code?

**Comment by fioriNTU** [comment]
@davidlange6 @slava77  Sorry to insist, but this is really crucial for us. Any news about it?

**Comment by slava77** [comment]
@fioriNTU 
I signed this PR a while back (2 weeks ago).
Nominally, all development release PRs (81X) get merged within a day or two after they are fully signed,
unless there are some issues.

@davidlange6 are there any issues with the changes made here?

**Comment by davidlange6** [comment]
yes, I don’t like the change to TrackerTopology.h as I have explained to @fioriNTU (exposing the internal implementation of the DetId packing). I have yet to trace through why it is needed for this PR.

**Comment by slava77** [comment]
On 5/17/16 6:23 AM, David Lange wrote:
> yes, I don’t like the change to TrackerTopology.h as I have explained to
> @fioriNTU (exposing the internal implementation of the DetId packing). I
> have yet to trace through why it is needed for this PR.
> 

OK, good to know.
I suggested to expose this after seeing the original implementation,
which was string based and rather DQM-specialized solution for something
that looked like "give me all you know about pixel subdetectors".
Probably another solution with more sensible interface from the
TrackerTopology can work out better.
It better be enum-based with string formatting done on the client code side.

	

>

**Comment by davidlange6** [comment]
> On May 17, 2016, at 4:09 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 5/17/16 6:23 AM, David Lange wrote:
> > yes, I don’t like the change to TrackerTopology.h as I have explained to
> > @fioriNTU (exposing the internal implementation of the DetId packing). I
> > have yet to trace through why it is needed for this PR.
> > 
> 
> OK, good to know.
> I suggested to expose this after seeing the original implementation,
> which was string based and rather DQM-specialized solution for something
> that looked like "give me all you know about pixel subdetectors".
> Probably another solution with more sensible interface from the
> TrackerTopology can work out better.
> It better be enum-based with string formatting done on the client code side.

TrackerTopology doesn’t tell you anything about the pixel sub detector actually included in the CMS geometry - but rather some maximal properties (maximum number of layers). I don’t see why that is interesting. It is instead the actual geometry that is interesting to monitor. anyway, I still haven’t looked at an example...

> 
> 
> 
> >

**Comment by davidlange6** [comment]
After more investigation I don't really see how the mask bit offsets are used here. Can you summarize? Contrary to what @slava77  says, this question was not due to his suggestion, as the code was just moved from one spot to another. @schneiml @fioriNTU

**Comment by schneiml** [comment]
The mask bits from TackerTopology are used to extract the respective qauntities from DetIds, as TT does. However, the interface to be used should not be based on method names, but allow
- Listing all quantities
- associating each quantity with a human-readable name
which is both not possible with TT. While the original plan was to provide TT with this functionality (which I'd still prefer), we moved the name-mapping to the PixelDQM-specific =GeometryInterface= since the user-facing names should be DQM-specific.

The information is in turn transformed into another representation (vector of functors), which is more specific and may contain bitmask-based as well as derived quantities.

**Comment by davidlange6** [comment]
what is a “name” in this context? can you give an example?
(and its not a geometry, so GeometryInterface should be renamed)

**Comment by schneiml** [comment]
Human-readable names are the strings visible around here: https://github.com/cms-sw/cmssw/pull/14230/files#diff-cf515bc8af8720e3dcf9d5d81ec7607fR201

These will later also be visible in the DQM, e.g. in the GUI. For this very reason I now think it is not useful to have them in TT, since we might want to change them later. However, I do think CMSSW might want to have a convention cast in code on how to label such things.

The `GeometryInterface` is meant to be an interface, as the name suggests, to all potentially geometry-dependent things. The fact that there is a lot of information about Topology encoded is due to the problem that there is no central place (to my knowledge) that has all the required information (esp. in an reasonably generic yet not too slow way). E.g. the Shell concept of DQM and the associated signed counting convention for ladders seems to be ad-hoc hardcoded in many places atm, so it has to be hardcoded in the `GeometryInterface` as well.

**Comment by davidlange6** [comment]
> On May 19, 2016, at 9:31 AM, Marcel Schneider <notifications@github.com> wrote:
> 
> Human-readable names are the strings visible around here: https://github.com/cms-sw/cmssw/pull/14230/files#diff-cf515bc8af8720e3dcf9d5d81ec7607fR201
> 

yes, I saw that, but then I didn’t manage to follow the actual usage - 

> These will later also be visible in the DQM, e.g. in the GUI. For this very reason I now think it is not useful to have them in TT, since we might want to change them later. However, I do think CMSSW might want to have a convention cast in code on how to label such things.
> 

pixels have some already, e.g.,
DataFormats/SiPixelDetId/interface/Pixel*Name*h


> The GeometryInterface is meant to be an interface, as the name suggests, to all potentially geometry-dependent things. The fact that there is a lot of information about Topology encoded is due to the problem that there is no central place (to my knowledge) that has all the required information (esp. in an reasonably generic yet not too slow way). E.g. the Shell concept of DQM and the associated signed counting convention for ladders seems to be ad-hoc hardcoded in many places atm, so it has to be hardcoded in the GeometryInterface as well.
> 

yes, its too bad that this is still the case as its not a new issue. There are some non-hard coded examples in pixels to follow such as DQM/SiPixelMonitorTrack/src/SiPixelTrackResidualSource.cc

**Comment by schneiml** [comment]
> pixels have some already, e.g.,
> DataFormats/SiPixelDetId/interface/Pixel*Name*h

Except that there are no actual "names" in there. There is some string
parsing, but I fail to see the corresponding formatting. Even if it were
available, this is not enough for the DQM, where we will have a folder
structure that contains some, but not all, of the fields.

> such as DQM/SiPixelMonitorTrack/src/SiPixelTrackResidualSource.cc

I don't know SiPixelTrackResidualSource by heart (yet), but I fail to
rememember almost anything *not* hardcoded there. I guess you are just
referring to the use of the classes mentioned above?

(looking more closely, `SiPixelTrackResidualSource.cc` is indeed not as
bad as other Pixel DQM plugins.)

**Comment by davidlange6** [comment]
@schneiml - what sort of string do you need? A human readable, physically meaningful, and unique name corresponding each DetId?

**Comment by schneiml** [comment]
> @schneiml - what sort of string do you need? A human readable, physically meaningful, and unique name corresponding each DetId?

No, I really want to see the structure. The DQM _might_ want to see all
of the "field" names and values ("Ladder_4", "Module_1", "Shell_pI") but
likely we don't want all of them but rather a subset. (e.g. Most often
we don't care about the module).

This is what the `namedPartitions` map allows: it lists all available
"fields", including the name and a way to acces the value of this field
(the information whether the field is applicable to some given DetId is
also important and ends to be encoded nowhere).

So, human readable and physically meaningful, but not necessarily
unique, since we care a lot about groups of DetIds (like full ladders,
full layers, etc.)

**Comment by davidlange6** [comment]
Ok - So I don't think anyone can provide whats needed without some specifications from DQM in that case.  However, what you need to do is to use the TrackerGeometry to derive the ranges of these things rather than the topology. [best would be to have a ESSource that depends on the geometry which provides a central location of the needed info, but there seems to be not so much interest in that old idea]

**Comment by schneiml** [comment]
Putting this stuff into an ESSource is a longer term goal, but we first have to figure out what we actually need (this is sort of understood now, but was not in the beginning) and also see what other users might be interested in. Until then we will have a local solution (maybe for ever). 

TrackerGeometry is used to figure out precise ranges, though that needs some reworking in the future - atm it e.g. does not represent the number of ladders _per layer_. But this does not matter for now, since this sort of information is (atm) only used in harvesting where ad-hoc counting is no problem. 

So the `GeometryInterface` will remain a facade around other things including TT, and the design space ranges from _encoding all relvant information there_ to having _only an interface_ to other sources (as originally intended). At the moment we have a compromise that uses the bitmasks from TT and counts from TG (and also Cabling) but encodes other things in place (Naming and Shells).

Are you ok with this solution?

**Comment by davidlange6** [comment]
@schneiml - I would find it reasonable to expose the maximum number of Pixel barrel (endcap, strip blah blah) layers (modules, other quantities, etc) in TrackerTopology via accessors returning unsigned ints, which sounds like it would be sufficient for your needs.

**Comment by schneiml** [comment]
@davidlange6 the maximum number is not really the main problem, since TrackerGeometry already provides this (not in a well readable form, but good enough that `GeometryInterface` can figure it out).

The main question I see is whether 
- we want to avoid changing TrackerTopology (which means a more ugly facade for me, since I need to wrap the methods in TT using an array of lambdas or sth. such that I can call them by-index or by-string-name, instead of by-methodname)
- or we want to fix the problem properly, which means providing a proper interface to access the fields of DetIds without using the different methods of TT (and probably more, to make it useful).

The third option is to leave things as they are, which is somewhere in the middle.

**Comment by davidlange6** [comment]
> On May 19, 2016, at 3:01 PM, Marcel Schneider <notifications@github.com> wrote:
> 
> @davidlange6 the maximum number is not really the main problem, since TrackerGeometry already provides this (not in a well readable form, but good enough that GeometryInterface can figure it out).
> 
> The main question I see is whether
> 
> 	• we want to avoid changing TrackerTopology (which means a more ugly facade for me, since I need to wrap the methods in TT using an array of lambdas or sth. such that I can call them by-index or by-string-name, instead of by-methodname)
> 	• or we want to fix the problem properly, which means providing a proper interface to access the fields of DetIds without using the different methods of TT (and probably more, to make it useful).
> The third option is to leave things as they are, which is somewhere in the middle.
> 

I have no problem with a better interface to the existing exposed information. (but that is not what I’ve seen in the PR either before or after @slava77 comments

**Comment by schneiml** [comment]
A better interface will take a considerable amount of extra work. The original idea of putting the human-readable names into TT as well turns out to be not really practicable. A variant with enum-keys might work, one might even find an appropriate enum somewhere in the code already. 

However, if we want to keep the bit masking all contained in TT (which seems like a good idea) and do not want to hand out functors (which sounds way to compilcated) we need to make sure that this is fast enough to be useful.

I could imagine a method 

     unsigned int getField(const DetId& id, SuitableEnum idx) { return ((id.rawId()>>startbit[idx])& mask[idx]);}

but this requires access to the bitmasks as an array. Something like this does not exist yet (or does it? serialization seems to have something in this direction but it sounds like we don't want to use that). 

This means a significant amount of planning and implementation.

**Comment by davidlange6** [comment]
@schneiml - I like the proposal, yes (possibly adding a sub detector argument as well).  adding a array of refs for each of the relevant fields should be straightforward

**Comment by schneiml** [comment]
Ok, I can try to implement that, however

- This will mean a new PR. The IB of this one long dead.
- Lots of rebasing to get everything up to date. I'd then add the other ~20 commits I have accumulated as well (a lot more files, more packages, more histograms, but nothing spectacular in there).

I'd go that path if we can get the new PR merged next week.  With another 30 day delay this is unacceptable. 

If you have any other suggestions, objections, renaming wishes, whatever, please tell me now.

**Comment by schneiml** [comment]
@davidlange6 I have now implemented something here:
[Data](https://github.com/schneiml/cmssw/blob/phase1pixeldqm-fourth/DataFormats/TrackerCommon/interface/TrackerTopology.h#L105) [Functions](https://github.com/schneiml/cmssw/blob/phase1pixeldqm-fourth/DataFormats/TrackerCommon/interface/TrackerTopology.h#L607)

Looks good to me, but causes ICE on GCC (4.9 as well as 5.1, clang accepts it), so we'll have to change it a bit (maybe moving the initializer to the .cc).

Please comment on whether this is what you expected.

**Comment by davidlange6** [comment]
@schneiml - yes, that looks good to me - i guess the pixel naming convention seems to be pf (or pxf) rather than pe for some reason.

**Comment by schneiml** [comment]
New PR #14586 is there. This one can probably be closed, if there are no issues with the new one.

Regarding the Endcap naming -- this is farly inconsistent, I defined the
difference to be that there is 2 PEs but only one P(X)F. But that does
not really make a difference in the given case, so I'll change to PF, to
be more in line with the rest of TT.

**Comment by davidlange6** [comment]
I agree that PE would seem more consistent..

**Comment by davidlange6** [comment]
I'll close this in favor of v4 - which hopefully @dmitrijus has a look at in the next days
